### PR TITLE
Feature/337 slf4j logging

### PR DESCRIPTION
**Pull request by digulla**
Notes:

- I didn't remove all calls to getLog() since some are used to set the flag "verbose". I don't know whether Maven will propagate the log level to all slf4j loggers. 
- For the same reason, I left the "if (isVerbose())" before the log calls but removed most `if (LOG.is...Level())`
- I kept the `if (LOG.is...Level())` is it protected a big code block.
- I replaced `"..." + variable` with `"... {}", variable` in all places I noticed. This should now be the default.
- I replaced `"... +" Arrays.asList(x)` where `x` is an array with `"{}", (Object)x`. Without the case, the Java compiler would try to copy the array elements into the varargs parameter of slf4j with the result that only the first element of the array would be logged.

### CODE REVIEW

File: src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java [hunk 1]
```diff
@@ -79,10 +76,7 @@ public void testRenderTemplateForThirdPartyFile()
         properties.put( "dependencyMap", licenseMap.toDependencyMap().entrySet() );
 
         String s = helper.renderTemplate( "/org/codehaus/mojo/license/third-party-file.ftl", properties );
-        if ( log.isInfoEnabled() )
-        {
-            log.info( s );
-        }
+        LOG.info( "{}", s );
```
Comment (ppalaga) [inline src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java]: If `s` is  a String, why not just `LOG.info(s)` ?

File: src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java [hunk 2]
```diff
@@ -118,10 +112,7 @@ public void testRenderTemplateForThirdPartyFileGroupByLicense()
 
         String s =
             helper.renderTemplate( "/org/codehaus/mojo/license/third-party-file-groupByLicense.ftl", properties );
-        if ( log.isInfoEnabled() )
-        {
-            log.info( s );
-        }
+        LOG.info( "{}", s );
```
Comment (ppalaga) [inline src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java]: If `s` is  a String, why not just `LOG.info(s)` ?

File: src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java [hunk 3]
```diff
@@ -149,9 +140,6 @@ public void testRenderTemplateForUpdateFileHeader()
         String s =
             helper.renderTemplate( "/org/codehaus/mojo/license/default-file-header-description.ftl", properties );
         Assert.assertEquals( "projectName\n$Id:$\n$HeadURL:$", s );
-        if ( log.isInfoEnabled() )
-        {
-            log.info( s );
-        }
+        LOG.info( "{}", s );
```
Comment (ppalaga) [inline src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java]: If `s` is  a String, why not just `LOG.info(s)` ?

File: pom.xml [hunk 4]
```diff
@@ -301,17 +301,22 @@
...
     </dependency>
 
     <dependency>
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>ch.qos.logback</groupId>
+      <artifactId>logback-classic</artifactId>
```
[the remark this answers is not in the document]
Comment (digulla) [reply pom.xml]: You saw that this is a just a test dependency?
I'm rather reluctant to turn this into an IT. As you can see, I have several tests and I will need more of those in the future because #201 would also need to check the log output.
If I turned all of them into ITs, it would take too much time for me to run them.
I could try to convert my library to work with the simple logging framework implementation used my Maven itself but I had the code and this solution was easiest for me.

File: src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java [hunk 5]
```diff
@@ -728,11 +727,8 @@ protected void init()
 
         long buildTimestamp = getBuildTimestamp();
 
-        if ( isVerbose() || getLog().isDebugEnabled() )
-        {
-            log.debug( "Build start   at : " + buildTimestamp );
-            log.debug( "third-party file : " + thirdPartyFile.lastModified() );
-        }
+        LOG.debug( "Build start   at : " + buildTimestamp );
+        LOG.debug( "third-party file : " + thirdPartyFile.lastModified() );
```
[the remark this answers is not in the document]
Comment (digulla) [reply src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java]: Done

File: src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java [hunk 6]
```diff
@@ -741,10 +737,7 @@ protected void init()
 
             File bundleFile = FileUtil.getFile( outputDirectory, bundleThirdPartyPath );
 
-            if ( isVerbose() || getLog().isDebugEnabled() )
-            {
-                log.debug( "bundle third-party file : " + bundleFile.lastModified() );
-            }
+            LOG.debug( "bundle third-party file : " + bundleFile.lastModified() );
```
[the remark this answers is not in the document]
Comment (digulla) [reply src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java]: Done

File: src/main/java/org/codehaus/mojo/license/AbstractFileHeaderMojo.java [hunk 7]
```diff
@@ -949,7 +947,7 @@ private void finalizeFile( File file, File processFile ) throws IOException
             }
             catch ( IOException e )
             {
-                getLog().warn( e.getMessage() );
+                LOG.warn( "Error updating {} -> {}", processFile, file, e );
```
[the remark this answers is not in the document]
Comment (digulla) [reply src/main/java/org/codehaus/mojo/license/AbstractFileHeaderMojo.java]: If the last argument is an exception, slf4j will render the stack trace. See https://www.slf4j.org/faq.html#paramException

File: src/main/java/org/codehaus/mojo/license/AggregatorAddThirdPartyMojo.java [hunk 8]
```diff
@@ -168,11 +171,9 @@ protected void init() throws Exception
     protected void doAction()
             throws Exception
     {
-        Log log = getLog();
-
         if ( isVerbose() )
         {
-            log.info( "After executing on " + reactorProjects.size() + " project(s)" );
+            LOG.info( "After executing on " + reactorProjects.size() + " project(s)" );
```
[the remark this answers is not in the document]
Comment (digulla) [reply src/main/java/org/codehaus/mojo/license/AggregatorAddThirdPartyMojo.java]: Done

File: src/main/java/org/codehaus/mojo/license/header/UpdateFileHeaderFilter.java [hunk 9]
```diff
@@ -86,14 +86,11 @@ protected FileHeader getNewHeader( FileHeader oldHeader )
...
-
             // can update copyright and it has changed
 
-            if ( log.isDebugEnabled() )
-            {
-                log.debug( "copyright has changed from [" + oldHeader.getCopyright() + "] to ["
-                        + newHeader.getCopyright() + "]" );
-            }
+            LOG.debug( "copyright has changed from [{}] to []",
+                    oldHeader.getCopyright(),
```
[the remark this answers is not in the document]
Comment (digulla) [reply src/main/java/org/codehaus/mojo/license/header/UpdateFileHeaderFilter.java]: Good catch!

File: src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java (hunk 1 above)
Comment (digulla) [reply src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java]: Because s could contain `{}`. It's rarely a problem except when you try something like `LOG.warn( s, e )`. If `s` contains `{}`, you get no stack trace anymore.
That's why I always use a format string. In the `buffer.toString()` case, it's more obvious because I save the "turn it into a string" when the logger is disabled.

File: src/main/java/org/codehaus/mojo/license/RemoveFileHeaderMojo.java [hunk 10]
```diff
@@ -595,7 +597,7 @@ private void finalizeFile( File file, File processFile ) throws IOException
             }
             catch ( IOException e )
             {
-                getLog().warn( e.getMessage() );
+                LOG.warn( e.getMessage(), e );
```
[the remark this answers is not in the document]
Comment (digulla) [reply src/main/java/org/codehaus/mojo/license/RemoveFileHeaderMojo.java]: I'm not sure whether that wasn't an omission or not. I'd expect Maven to suppress the stack trace unless it's being run with -X.
That said, my general rule is always to include the stack trace when handling exceptions. I've been burned once too many by code what just gave me the message but no stack trace.

**Comment by digulla** [comment]
The last two commits fix a deprecation and a warning I got in the POM. Since they are small changes, I didn't want to create a PR for them.

File: pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (digulla) [reply pom.xml]: Note: I'm trying to get this into the slf4j code base: https://jira.qos.ch/browse/SLF4J-461

[the remark this answers is not in the document]
Comment (digulla) [reply pom.xml]: I spent a couple of hours to write a very simple logger implementation which would also make the tests pass. The code is here: https://github.com/digulla/license-maven-plugin/commit/5eb32342cd1bd608aec98a9bc18cba303fe6d1a9
It quite a few lines of code. On the plus side, we can now add additional checks (= number of {} matches number of arguments).

File: src/main/java/org/codehaus/mojo/license/AbstractFileHeaderMojo.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ppalaga) [reply src/main/java/org/codehaus/mojo/license/AbstractFileHeaderMojo.java]: Do we really want to see the stack trace here given that we were not showing it in the past?

File: src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java (hunk 1 above)
Comment (ppalaga) [reply src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java]: Why would s containing `{}` be a problem? With simple logger, the string is passed as is no matter what it contains: https://github.com/qos-ch/slf4j/blob/master/slf4j-simple/src/main/java/org/slf4j/simple/SimpleLogger.java#L479-L481 and there is no varargs in `Logger.info(String)`. `Logger.info(String)` is the simplest placeholder-less variant.

Comment (digulla) [reply src/test/java/org/codehaus/mojo/license/api/FreeMarkerHelperTest.java]: Because in some corner cases, the logging API behaves differently. It works in this case but if you try that with a logger method that takes more arguments, something weird can happen. I prefer to use the same pattern everywhere, so I'm not getting surprised.
My fear is mostly that someone with less knowledge will modify the code later on.

File: pom.xml [hunk 11]
```diff
@@ -137,6 +137,12 @@
 
   <dependencyManagement>
     <dependencies>
+      <dependency>
+        <groupId>org.apache.maven.plugin-tools</groupId>
+        <artifactId>maven-plugin-annotations</artifactId>
+        <version>${pluginPluginVersion}</version>
+        <scope>provided</scope>
```
Comment (ppalaga) [inline pom.xml]: Please do not use `<scope>` under dependencyManagement. Having it there is not reliable and thus redundant.

File: pom.xml [hunk 12]
```diff
@@ -147,6 +153,17 @@
...
+      <dependency>
+        <groupId>org.slf4j</groupId>
+        <artifactId>slf4j-api</artifactId>
+        <version>1.7.26</version>
+      </dependency>
+      <dependency>
+        <groupId>ch.qos.logback</groupId>
+        <artifactId>logback-classic</artifactId>
+        <version>1.2.3</version>
+        <scope>test</scope>
```
Comment (ppalaga) [inline pom.xml]: Please do not use `<scope>` under dependencyManagement. Having it there is not reliable and thus redundant.

File: pom.xml (hunk 11 above)
Comment (digulla) [reply pom.xml]: I can remove it but I think without it, some dependency can then force  it into the scope compiled. As it is, even no scope is given at other places, it should be in `provided` in the final classpath.
Am I wrong?

Comment (digulla) [reply pom.xml]: SO seems to support my approach: https://stackoverflow.com/questions/15221299/dependencymanagement-and-scope

File: pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ppalaga) [reply pom.xml]: Yes, I saw, this is a test dependency and that `org.codehaus.mojo.license.logback.*` is test code and I still think it is too much code for what it does. I'd prefer to stop checking the log output for the cases we have now. The methods under test would be tested well enough by checking their return values or thrown exceptions.

If you insist on some way to verify logged messages in the future, I prefer discussing it separately at the point in time when we start needing it. 

This PR contains a lot of useful changes and I would not like to block it because of this.

**Comment by ppalaga** [comment]
Summary: two last things open from my PoV:

1. `org.codehaus.mojo.license.logback.*` removal

2. `<scope>` in depManagement removal.

File: pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (digulla) [reply pom.xml]: Maven communicates with users by the text we print on the console. In other software, logging is used to do post mortem debugging. But even there, I'd argue you want to include important log messages in the tests to make sure they are logged and that they contain useful information.
That's why I argue the things we log are part of the API contract. It's part of the "return values" which you want me to test.
Therefore, I want to include those in the tests. I did that before and it wasn't a problem.

Many people don't know how to test this. That's why there is a feeling "not possible" / "don't try this".
I do this in all my projects. The code I added is a simplified copy of my own log testing library. The code was written two years ago and I had no bugs, just a couple of feature requests in that time.

There are several issues in the license-maven-plugin bugtracker which ask to make changes to the output. Without the framework, we won't be able to test those changes in unit tests.
I don't like ITs. They are slow, brittle, take a long time to set up, hard to debug, and when they break, you have to spend a lot of time figuring out what they are supposed to do. To add insult to injury, each of the ITs in the project take 30-120 seconds to run on Windows.

As for "we don't need it now": Those tests need it now. They already existed. They checked log output before and if I removed that, I would cripple existing tests. I wanted to avoid that.

Therefore, I think I have several good arguments for my approach.

What exactly is the problem? Is it too many lines of code per PR? Or are you afraid that some change will break the code and no one will be able to fix it since it's too complicated?

[the remark this answers is not in the document]
Comment (ppalaga) [reply pom.xml]: Yes, I agree ITs are often an overkill (and perhaps also an overkill for this particular case) At the same time I dare to insist that adding the logback dependency and 300+ lines of new logback code is an overkill too. I am not doubting the quality of the code, I have concerns about the future maintainability of the project. Less code and less dependencies are better. Logback and five new classes are too much for what we need.

As I said, I'd be ready to sacrifice the log verification altogether and you are right that I do not consider log messages to be a part of the method contract. I'd say if the messages are to be seen as part of the contract, they should become part of the directly testable method output. Either passing a `org.slf4j.Logger` as a method param or some simplified easily mockable interface would be IMO much cheaper than logback + 300 LoC. I'd still prefer to split this discussion from the present PR, so that other potentially conflicting work can go on.

File: pom.xml (hunk 11 above)
Comment (ppalaga) [reply pom.xml]: I frankly do not know what happens if `maven-plugin-annotations` accidentally get into `compile` scope. Having `<scope>provided</scope>` only in the dependencies section worked so far. So let's keep the old strategy and remove the scopes from the dependencyManagement.

**Comment by digulla** [comment]
I'm now running out of time. I'll be at a conference next week, see you in July.

File: pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (digulla) [reply pom.xml]: I'm sorry that I can't do this with less code. I strongly believe that the log messages should be checked which is why I invented a framework to be able to do it in a comfortable and safe way.
The slf4j API isn't very well suited for this, which is why I need that much code.
I also don't want to cripple existing tests and I'm out of time.
I'm really trying here to help but I feel that my expertise isn't appreciated. I'm doing this to help and to scratch an itch that I have and at the moment, it's not fun at all. If you don't accept this PR either, I've wasted some 30 hours which I could have put into a fork that fixes my problems. So for me, this isn't really working out and I have a feeling it's not working for you either.

**Comment by ppalaga** [comment]
@digulla it is a pitty. This is not the first time that reviewing your large PR costs me time with no result. Please send PRs that have chance to get reviewed and merged quickly. Please discuss potentially controversial things like adding new dependencies upfront. To move forward with the next release, unless you veto within a reasonable time, I am going to make a new PR out of the acceptable parts of this PR.

**Comment by ppalaga** [comment]
Replaced by https://github.com/mojohaus/license-maven-plugin/pull/342
